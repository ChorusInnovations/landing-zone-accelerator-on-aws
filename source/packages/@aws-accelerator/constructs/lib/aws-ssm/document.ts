/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
const path = require('path');

export interface IDocument extends cdk.IResource {
  readonly documentName: string;
}

export interface DocumentProps {
  readonly name: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly content: any | cdk.IResolvable;
  readonly documentType: string;
  readonly sharedWithAccountIds: string[];
}

export class Document extends cdk.Resource implements IDocument {
  readonly documentName: string;

  constructor(scope: Construct, id: string, props: DocumentProps) {
    super(scope, id);

    const resource = new cdk.aws_ssm.CfnDocument(this, 'Resource', {
      name: props.name,
      content: props.content,
      documentType: props.documentType,
    });

    this.documentName = resource.ref;

    // Also need a custom resource to do the share
    if (props.sharedWithAccountIds.length > 0) {
      const SHARE_SSM_DOCUMENT = 'Custom::SSMShareDocument';

      const cr = cdk.CustomResourceProvider.getOrCreateProvider(this, SHARE_SSM_DOCUMENT, {
        codeDirectory: path.join(__dirname, 'share-document/dist'),
        runtime: cdk.CustomResourceProviderRuntime.NODEJS_14_X,
        policyStatements: [
          {
            Sid: 'ShareDocumentActions',
            Effect: 'Allow',
            Action: ['ssm:DescribeDocumentPermission', 'ssm:ModifyDocumentPermission'],
            Resource: cdk.Stack.of(this).formatArn({
              service: 'ssm',
              resource: 'document',
              arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
              resourceName: '*',
            }),
          },
        ],
      });

      new cdk.CustomResource(this, 'ShareDocument', {
        resourceType: SHARE_SSM_DOCUMENT,
        serviceToken: cr.serviceToken,
        properties: {
          name: this.documentName,
          accountIds: props.sharedWithAccountIds,
        },
      });
    }
  }
}
