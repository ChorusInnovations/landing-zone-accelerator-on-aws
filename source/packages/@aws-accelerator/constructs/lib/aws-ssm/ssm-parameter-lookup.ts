/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * SsmParameterLookupProps
 */
export interface SsmParameterLookupProps {
  /**
   * Name of the parameter
   */
  readonly name: string;
  /**
   * Parameter account id
   */
  readonly accountId: string;
  /**
   * Custom resource lambda log group encryption key
   */
  readonly kmsKey?: cdk.aws_kms.Key;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays?: number;
}

const CROSS_ACCOUNT_ACCESS_ROLE_NAME = 'AWSAccelerator-CrossAccount-SsmParameter-Role';

/**
 * SsmParameterLookup class - to get ssm parameter value from other account
 */
export class SsmParameterLookup extends Construct {
  public readonly value: string = '';

  static isLogGroupConfigured = false;

  constructor(scope: Construct, id: string, props: SsmParameterLookupProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::SsmGetParameterValue';

    const roleArn = `arn:${cdk.Stack.of(this).partition}:iam::${
      props.accountId
    }:role/${CROSS_ACCOUNT_ACCESS_ROLE_NAME}`;

    const customResourceLambdaFunction = new cdk.aws_lambda.Function(this, 'SsmGetParameterValueFunction', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'get-param-value/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      description: `Custom resource provider to get ssm parameter ${props.name} value`,
    });

    /**
     * Pre-Creating log group to enable encryption and log retention.
     * Below construct needs to be static
     * isLogGroupConfigured flag used to make sure log group construct synthesize only once in the stack
     */
    if (!SsmParameterLookup.isLogGroupConfigured) {
      const logGroup = new cdk.aws_logs.LogGroup(this, 'LogGroup', {
        logGroupName: `/aws/lambda/${customResourceLambdaFunction.functionName}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });

      logGroup.node.addDependency(customResourceLambdaFunction);

      // Enable the flag to indicate log group configured
      SsmParameterLookup.isLogGroupConfigured = true;
    }

    const policyStatements: cdk.aws_iam.PolicyStatement[] = [];

    policyStatements.push(
      new cdk.aws_iam.PolicyStatement({
        sid: 'SsmGetParameterActions',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['ssm:GetParameters', 'ssm:GetParameter', 'ssm:DescribeParameters'],
        resources: ['*'],
      }),
      new cdk.aws_iam.PolicyStatement({
        sid: 'StsAssumeRoleActions',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['sts:AssumeRole'],
        resources: [roleArn],
      }),
    );

    policyStatements.forEach(policyStatement => {
      customResourceLambdaFunction?.addToRolePolicy(policyStatement);
    });

    const customResourceProvider = new cdk.custom_resources.Provider(this, 'CustomResourceProvider', {
      onEventHandler: customResourceLambdaFunction,
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: customResourceProvider.serviceToken,
      properties: {
        region: cdk.Stack.of(this).region,
        parameterAccountID: props.accountId,
        parameterName: props.name,
        assumeRoleArn: roleArn,
        invokingAccountID: cdk.Stack.of(this).account,
      },
    });

    this.value = resource.ref;
  }
}
