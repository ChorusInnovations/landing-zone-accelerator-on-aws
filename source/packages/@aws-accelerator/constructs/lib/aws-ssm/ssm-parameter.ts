/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

export enum SsmParameterType {
  GET = 'GET',
  PUT = 'PUT',
}
/**
 * SsmParameterProps
 */
export interface SsmParameterProps {
  readonly region: string;
  readonly partition: string;
  /**
   * SSM Parameter
   */
  readonly parameter: {
    /**
     * Name of the parameter
     */
    name: string;
    /**
     * Target account id of the parameter
     */
    accountId: string;
    /**
     * Role name to assume to access the parameter in target account
     */
    roleName: string;
    /**
     * Optional value to put when when using SsmParameterType.PUT
     */
    value?: string;
  };
  readonly invokingAccountID: string;

  /**
   * The type of SSM parameter operation
   */
  readonly type: SsmParameterType;
}

/**
 * SsmParameter class - to get ssm parameter value from other account
 */
export class SsmParameter extends Construct {
  public readonly parameterName: string;
  public readonly value: string = '';

  constructor(scope: Construct, id: string, props: SsmParameterProps) {
    super(scope, id);

    this.parameterName = props.parameter.name;
    const assumeRoleArn = `arn:${props.partition}:iam::${props.parameter.accountId}:role/${props.parameter.roleName}`;

    let RESOURCE_TYPE: string;
    let codeDir: string;
    let desc: string;
    let logicalId: string;
    const policyStatements: cdk.aws_iam.PolicyStatement[] = [];

    if (props.type === SsmParameterType.GET) {
      RESOURCE_TYPE = 'Custom::SsmGetParameterValue';
      codeDir = 'get-param-value/dist';
      desc = `Custom resource provider to get ssm parameter ${props.parameter.name} value`;
      logicalId = 'SsmGetParameterValueFunction';

      // Push policy statement
      policyStatements.push(
        new cdk.aws_iam.PolicyStatement({
          sid: 'SsmGetParameterActions',
          effect: cdk.aws_iam.Effect.ALLOW,
          actions: ['ssm:GetParameters', 'ssm:GetParameter', 'ssm:DescribeParameters'],
          resources: ['*'],
        }),
      );
    } else if (props.type === SsmParameterType.PUT) {
      // Check if parameter value is included in props
      if (!props.parameter.value) {
        throw new Error('parameter.value property required when type is set to PUT');
      }

      RESOURCE_TYPE = 'Custom::SsmPutParameterValue';
      codeDir = 'put-param-value/dist';
      desc = `Custom resource provider to put ssm parameter ${props.parameter.name} value`;
      logicalId = 'SsmPutParameterValueFunction';

      // Push policy statement
      policyStatements.push(
        new cdk.aws_iam.PolicyStatement({
          sid: 'SsmPutParameterActions',
          effect: cdk.aws_iam.Effect.ALLOW,
          actions: ['ssm:DeleteParameter', 'ssm:PutParameter'],
          resources: ['*'],
        }),
      );

      this.value = props.parameter.value;
    } else {
      throw new Error(`SSM parameter type ${props.type} is invalid`);
    }

    policyStatements.push(
      new cdk.aws_iam.PolicyStatement({
        sid: 'StsAssumeRoleActions',
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['sts:AssumeRole'],
        resources: [assumeRoleArn],
      }),
    );

    // Create custom resource
    const customResourceLambdaFunction = new cdk.aws_lambda.Function(this, logicalId, {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, codeDir)),
      runtime: cdk.aws_lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      description: desc,
    });

    policyStatements.forEach(policyStatement => {
      customResourceLambdaFunction?.addToRolePolicy(policyStatement);
    });

    const customResourceProvider = new cdk.custom_resources.Provider(this, 'CustomResourceProvider', {
      onEventHandler: customResourceLambdaFunction,
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: customResourceProvider.serviceToken,
      properties: {
        region: props.region,
        parameterAccountID: props.parameter.accountId,
        parameterName: props.parameter.name,
        parameterValue: props.parameter.value,
        assumeRoleArn: assumeRoleArn,
        invokingAccountID: props.invokingAccountID,
      },
    });

    if (props.type === SsmParameterType.GET) {
      this.value = resource.ref;
    }
  }
}
