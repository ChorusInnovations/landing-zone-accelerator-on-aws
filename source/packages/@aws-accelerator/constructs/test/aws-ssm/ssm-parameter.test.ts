/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { SsmParameter, SsmParameterType } from '../../lib/aws-ssm/ssm-parameter';

const testNamePrefix = 'Construct(SsmParameter): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new SsmParameter(stack, 'SsmParameter', {
  region: 'us-east-1',
  partition: 'aws',
  parameter: {
    name: `/accelerator/network/vpcPeering/name/id`,
    accountId: '111111111111',
    roleName: `AWSAccelerator-VpcPeeringRole-222222222222`,
    value: 'vp-123123123',
  },
  invokingAccountID: '333333333333',
  type: SsmParameterType.PUT,
});

/**
 * SsmParameter construct test
 */
describe('SsmParameter', () => {
  /**
   * Number of Lambda function resource test
   */
  test(`${testNamePrefix} Lambda function resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::Lambda::Function', 2);
  });
});
