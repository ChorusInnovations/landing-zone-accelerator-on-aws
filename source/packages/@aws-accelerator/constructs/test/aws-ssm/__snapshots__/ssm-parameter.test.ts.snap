// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SsmParameter Construct(SsmParameter):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "SsmParameter39B3125C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "SsmParameterCustomResourceProviderframeworkonEventE9ED19F6",
            "Arn",
          ],
        },
        "assumeRoleArn": "arn:aws:iam::111111111111:role/AWSAccelerator-VpcPeeringRole-222222222222",
        "invokingAccountID": "333333333333",
        "parameterAccountID": "111111111111",
        "parameterName": "/accelerator/network/vpcPeering/name/id",
        "parameterValue": "vp-123123123",
        "region": "us-east-1",
      },
      "Type": "Custom::SsmPutParameterValue",
      "UpdateReplacePolicy": "Delete",
    },
    "SsmParameterCustomResourceProviderframeworkonEventE9ED19F6": Object {
      "DependsOn": Array [
        "SsmParameterCustomResourceProviderframeworkonEventServiceRoleDefaultPolicyF43DF488",
        "SsmParameterCustomResourceProviderframeworkonEventServiceRoleE1067FDA",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/SsmParameter/CustomResourceProvider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "SsmParameterSsmPutParameterValueFunction190E66D1",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "SsmParameterCustomResourceProviderframeworkonEventServiceRoleE1067FDA",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SsmParameterCustomResourceProviderframeworkonEventServiceRoleDefaultPolicyF43DF488": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SsmParameterSsmPutParameterValueFunction190E66D1",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SsmParameterSsmPutParameterValueFunction190E66D1",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SsmParameterCustomResourceProviderframeworkonEventServiceRoleDefaultPolicyF43DF488",
        "Roles": Array [
          Object {
            "Ref": "SsmParameterCustomResourceProviderframeworkonEventServiceRoleE1067FDA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SsmParameterCustomResourceProviderframeworkonEventServiceRoleE1067FDA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SsmParameterSsmPutParameterValueFunction190E66D1": Object {
      "DependsOn": Array [
        "SsmParameterSsmPutParameterValueFunctionServiceRoleDefaultPolicyF7CBFD89",
        "SsmParameterSsmPutParameterValueFunctionServiceRoleA6975135",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Custom resource provider to put ssm parameter /accelerator/network/vpcPeering/name/id value",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "SsmParameterSsmPutParameterValueFunctionServiceRoleA6975135",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "SsmParameterSsmPutParameterValueFunctionServiceRoleA6975135": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SsmParameterSsmPutParameterValueFunctionServiceRoleDefaultPolicyF7CBFD89": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssm:DeleteParameter",
                "ssm:PutParameter",
              ],
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "SsmPutParameterActions",
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::111111111111:role/AWSAccelerator-VpcPeeringRole-222222222222",
              "Sid": "StsAssumeRoleActions",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SsmParameterSsmPutParameterValueFunctionServiceRoleDefaultPolicyF7CBFD89",
        "Roles": Array [
          Object {
            "Ref": "SsmParameterSsmPutParameterValueFunctionServiceRoleA6975135",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
